### ISSUE TITLE
Iterator: index out of range

### ISSUE THREAD

**Issue by tangtony** [allegro/bigcache#107]
Has anyone else seen this issue when using the iterator? It only happens occasionally and I haven't found a way to reproduce this:
```sh
panic: runtime error: index out of range

goroutine 186467 [running]:
encoding/binary.littleEndian.Uint64(...)
        /usr/local/go/src/encoding/binary/binary.go:76
myapp/vendor/github.com/allegro/bigcache.readTimestampFromEntry(...)
        /go/src/myapp/vendor/github.com/allegro/bigcache/encoding.go:45
myapp/vendor/github.com/allegro/bigcache.(*EntryInfoIterator).Value(0xc024444050, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/myapp/vendor/github.com/allegro/bigcache/iterator.go:117 +0x555
myapp/cache.GetAllEntries.func1(0xc004b3d020)
        /go/src/myapp/cache/cache.go:41 +0x9f
created by myapp/cache.GetAllEntries
        /go/src/myapp/cache/cache.go:38 +0x58
```

My code looks like this:
```go
func GetAllEntries() <-chan []byte {
	entries := make(chan []byte)
	go func() {
		iter := cache.Iterator()
		for iter.SetNext() {
			if entry, err := iter.Value(); err == nil {
				entries <- entry.Value()
			}
		}
		close(entries)
	}()
	return entries
}
```

**Comment by cristaloleg** [comment]
Hm.. that's interesting.

**Comment by un000** [comment]
```

**Comment by un000** [comment]
```
package bigcache_test

import (
	"log"
	"runtime"
	"strconv"
	"testing"
	"time"

	"github.com/allegro/bigcache"
)

func TestBigCacheIter(t *testing.T) {
	bc, err := bigcache.NewBigCache(bigcache.DefaultConfig(1 *time.Minute))
	if err != nil {
		log.Panic(err)
	}

	go func() {
		for i := 0; i < 10000; i++ {
			err := bc.Set(strconv.Itoa(i), []byte("aaaaaaa"))
			if err != nil {
				log.Panic(err)
			}

			runtime.Gosched()
		}
	}()

	for i := 0; i < 10000; i++ {
		iter := bc.Iterator()
		for iter.SetNext() {
			v, err := iter.Value()
			if err == nil {
				log.Panic(err)
			}

			_ = v.Value()
		}
	}
}
```

**Comment by un000** [comment]
```
package bigcache_test

import (
	"log"
	"runtime"
	"strconv"
	"testing"
	"time"

	"github.com/allegro/bigcache"
)

func TestBigCacheIter(t *testing.T) {
	bc, err := bigcache.NewBigCache(bigcache.DefaultConfig(1 *time.Minute))
	if err != nil {
		log.Panic(err)
	}

	go func() {
		for i := 0; i < 10000; i++ {
			err := bc.Set(strconv.Itoa(i), []byte("aaaaaaa"))
			if err != nil {
				log.Panic(err)
			}

			runtime.Gosched()
		}
	}()

	for i := 0; i < 10000; i++ {
		iter := bc.Iterator()
		for iter.SetNext() {
			v, err := iter.Value()
			if err == nil {
				log.Panic(err)
			}

			_ = v.Value()
		}
	}
}

```


```
GOROOT=/usr/local/Cellar/go/1.11/libexec #gosetup
GOPATH=/Users/un0/go #gosetup
/usr/local/Cellar/go/1.11/libexec/bin/go test -c -race -o /private/var/folders/ld/fvrvwzgj7zj9n15qpftx_tw40000gn/T/___cache_test_go github.com/un000/test/bigcache #gosetup
/usr/local/Cellar/go/1.11/libexec/bin/go tool test2json -t /private/var/folders/ld/fvrvwzgj7zj9n15qpftx_tw40000gn/T/___cache_test_go -test.v -test.run ^TestBigCacheIter$ #gosetup
=== RUN   TestBigCacheIter
==================
WARNING: DATA RACE
Read at 0x00c006115e00 by goroutine 6:
  github.com/allegro/bigcache.(*cacheShard).copyKeys()
      /Users/un0/go/src/github.com/allegro/bigcache/shard.go:147 +0x74
  github.com/allegro/bigcache.(*EntryInfoIterator).SetNext()
      /Users/un0/go/src/github.com/allegro/bigcache/iterator.go:72 +0x1f8
  github.com/un000/test/bigcache_test.TestBigCacheIter()
      /Users/un0/go/src/github.com/un000/test/bigcache/cache_test.go:32 +0x2e3
  testing.tRunner()
      /usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:827 +0x162

Previous write at 0x00c006115e00 by goroutine 11:
  runtime.mapassign_fast64()
      /usr/local/Cellar/go/1.11/libexec/src/runtime/map_fast64.go:92 +0x0
  github.com/allegro/bigcache.(*cacheShard).set()
      /Users/un0/go/src/github.com/allegro/bigcache/shard.go:76 +0x302
  github.com/allegro/bigcache.(*BigCache).Set()
      /Users/un0/go/src/github.com/allegro/bigcache/bigcache.go:117 +0x156
  github.com/un000/test/bigcache_test.TestBigCacheIter.func1()
      /Users/un0/go/src/github.com/un000/test/bigcache/cache_test.go:21 +0xcc

Goroutine 6 (running) created at:
  testing.(*T).Run()
      /usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:878 +0x650
  testing.runTests.func1()
      /usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:1119 +0xa8
  testing.tRunner()
      /usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:827 +0x162
  testing.runTests()
      /usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:1117 +0x4ee
  testing.(*M).Run()
      /usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:1034 +0x2ee
  main.main()
      _testmain.go:42 +0x221

Goroutine 11 (running) created at:
  github.com/un000/test/bigcache_test.TestBigCacheIter()
      /Users/un0/go/src/github.com/un000/test/bigcache/cache_test.go:19 +0x2cb
  testing.tRunner()
      /usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:827 +0x162
==================
2018/11/25 13:04:52 <nil>
--- FAIL: TestBigCacheIter (0.18s)
    testing.go:771: race detected during execution of test
panic: <nil> [recovered]
	panic: <nil>

goroutine 19 [running]:
testing.tRunner.func1(0xc0000da100)
	/usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:792 +0x6a7
panic(0x11b1b80, 0xc0131e4000)
	/usr/local/Cellar/go/1.11/libexec/src/runtime/panic.go:513 +0x1b9
log.Panic(0xc0000bddc0, 0x1, 0x1)
	/usr/local/Cellar/go/1.11/libexec/src/log/log.go:326 +0xe8
github.com/un000/test/bigcache_test.TestBigCacheIter(0xc0000da100)
	/Users/un0/go/src/github.com/un000/test/bigcache/cache_test.go:35 +0x3b1
testing.tRunner(0xc0000da100, 0x11ecda0)
	/usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:827 +0x163
created by testing.(*T).Run
	/usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:878 +0x651

Process finished with exit code 1
```

**Comment by un000** [comment]
https://github.com/allegro/bigcache/pull/108

**Comment by mxplusb** [comment]
Closing as resolved in #108.
